drivers/fpga/fpga_altera.c

```diff
@@ -0,0 +1,463 @@
+	k_sem_init(&(priv_data.smc_sem), 0, 1);
+
+	smc_cmd[SMC_REQUEST_A2_INDEX] = FIELD_GET(BIT(0), enable);
+
+	smc_cmd[SMC_REQUEST_A2_INDEX] |= BIT(1);
+	smc_cmd[SMC_REQUEST_A3_INDEX] = BRIDGE_MASK;
+
+	ret = smc_send(SIP_SVC_PROTO_CMD_SYNC, SMC_FUNC_ID_SET_HPS_BRIDGES, smc_cmd, &priv_data);
+	if (ret) {
+		LOG_ERR("Failed to send the smc Command !!");
+		return ret;
+	}
+
+	/* Wait for the SiP SVC callback */
+	k_sem_take(&(priv_data.smc_sem), K_FOREVER);
+
+	/* Check error code */
+	if (priv_data.response.a0) {
+		ret = -ENOMSG;
+	}
+
+	return ret;
+}
+
+static int altera_fpga_on(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	int32_t ret = 0;
+
+	/* Opening SIP SVC session */
+	ret = svc_client_open();
+	if (ret) {
+		LOG_ERR("Client open Failed!");
+		return ret;
+	}
+
+	/* Check FPGA status before bridge enable/disable */
+	ret = is_fpga_config_not_ready();
+	if (ret) {
+		LOG_ERR("FPGA not ready. Bridge reset aborted!");
+		svc_client_close();
+		return -EIO;
+	}
+
+	/* Bridge reset */
+	ret = socfpga_bridges_reset(0x01);
+	if (ret) {
+		LOG_ERR("Bridge reset failed");
+	}
+
+	/* Ignoring the return value to return bridge reset status */
+	if (svc_client_close()) {
+		LOG_ERR("Unregistering & Closing failed");
+	}
+
+	return ret;
+}
+
+static int altera_fpga_off(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	int32_t ret = 0;
+
+	/* Opening SIP SVC session */
+	ret = svc_client_open();
+	if (ret) {
+		LOG_ERR("Client open Failed!");
+		return ret;
+	}
+
+	/* Check FPGA status before bridge enable/disable */
+	ret = is_fpga_config_not_ready();
+	if (ret) {
+		LOG_ERR("FPGA not ready. Bridge reset aborted!");
+		svc_client_close();
+		return -EIO;
+	}
+
+	/* Bridge reset */
+	ret = socfpga_bridges_reset(0x00);
+	if (ret) {
+		LOG_ERR("Bridge reset failed");
+	}
+
+	/* Ignoring the return value to return bridge reset status */
+	if (svc_client_close()) {
+		LOG_ERR("Unregistering & Closing failed");
+	}
+
+	return ret;
+}
+
+static int altera_fpga_init(const struct device *dev)
+{
+	if (!dev) {
+		LOG_ERR("No such device found");
+		return -ENODEV;
+	}
+
+	mailbox_smc_dev = sip_svc_get_controller("smc");
+	if (!mailbox_smc_dev) {
+		LOG_ERR("Arm SiP service not found");
+		return -ENODEV;
+	}
+
+	mailbox_client_token = sip_svc_register(mailbox_smc_dev, NULL);
+	if (mailbox_client_token == SIP_SVC_ID_INVALID) {
+		mailbox_smc_dev = NULL;
+		LOG_ERR("Mailbox client register fail");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int altera_fpga_reset(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+	return -ENOTSUP;
+}
+
+static int altera_fpga_load(const struct device *dev, uint32_t *image_ptr, uint32_t img_size)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(image_ptr);
+	ARG_UNUSED(img_size);
+
+	return -ENOTSUP;
+}
+
+static enum FPGA_status altera_fpga_get_status(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+	return -ENOTSUP;
+}
+
+static const char *altera_fpga_get_info(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+	return NULL;
```

Comment:
I am afraid it still won't work, the FPGA layer is too limited at this point.

You'll get a NULL here: https://github.com/zephyrproject-rtos/zephyr/blob/9b17a952f6df812fad42e124e2b70eea89c77ff5/include/zephyr/drivers/fpga.h#L126-L132

and it will propagate here:
https://github.com/zephyrproject-rtos/zephyr/blob/9b17a952f6df812fad42e124e2b70eea89c77ff5/drivers/fpga/fpga_shell.c#L127

It is similar for other functions as well.

Reply:
@tgorochowik  Agree, NULL will be propagated to fpga_shell but we can't return -ENOSUP from the common layer i.e. fpga.h as the return type for the fpga_get_info function is of type "char *" or you are suggestion is to return "-ENOSUP" as string.
Either way, we have to return NULL or "-ENOSUP" either from the driver or from the common layer.

Reply:
Is there a significant reason to return `NULL` or `"-ENOTSUP"` here?

For lack of a better option, you could return e.g. the crc of the bitstream if it is loaded, otherwise, an empty string.

https://github.com/zephyrproject-rtos/zephyr/blob/e3ff9930003cfbbaf1409d4aee51e39057358b0d/drivers/fpga/fpga_ice40.c#L495

Reply:
Thanks @cfriedt for you valuable comment I agree with you we can return the empty string in this case